File: WinPaletter/LogonUI/LogonUIXP.vb
Imports System.ComponentModel

Public Class LogonUIXP
    Private Sub LogonUIXP_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        LoadLanguage
        ApplyStyle(Me)
        Icon = LogonUI.Icon
        Button12.Image = MainFrm.Button20.Image.Resize(16, 16)
        ApplyFromCP(My.CP)
    End Sub

    Protected Overrides Sub OnDragOver(e As DragEventArgs)
        If TypeOf e.Data.GetData("WinPaletter.UI.Controllers.ColorItem") Is UI.Controllers.ColorItem Then
            Focus()
            BringToFront()
        Else
            Exit Sub
        End If

        MyBase.OnDragOver(e)
    End Sub

    Sub ApplyFromCP(CP As CP)
        With CP.LogonUIXP
            Toggle1.Checked = .Enabled
            Select Case .Mode
                Case CP.Structures.LogonUIXP.Modes.Default
                    RadioImage1.Checked = True
                Case CP.Structures.LogonUIXP.Modes.Win2000
                    RadioImage2.Checked = True
                Case Else
                    RadioImage1.Checked = True
            End Select
            color_pick.BackColor = .BackColor
            CheckBox1.Checked = .ShowMoreOptions
        End With
    End Sub

    Sub ApplyToCP(CP As CP)
        With CP.LogonUIXP
            .Enabled = Toggle1.Checked
            If RadioImage1.Checked Then .Mode = CP.Structures.LogonUIXP.Modes.Default Else .Mode = CP.Structures.LogonUIXP.Modes.Win2000
            .BackColor = color_pick.BackColor
            .ShowMoreOptions = CheckBox1.Checked
        End With
    End Sub

    Private Sub Button11_Click(sender As Object, e As EventArgs) Handles Button11.Click
        If OpenFileDialog1.ShowDialog = DialogResult.OK Then
            Dim CPx As New CP(CP.CP_Type.File, OpenFileDialog1.FileName)
            ApplyFromCP(CPx)
            CPx.Dispose()
        End If
    End Sub

    Private Sub Button9_Click(sender As Object, e As EventArgs) Handles Button9.Click
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyFromCP(CPx)
        CPx.Dispose()
    End Sub

    Private Sub Button12_Click(sender As Object, e As EventArgs) Handles Button12.Click
        Using _Def As CP = CP_Defaults.From(My.PreviewStyle)
            ApplyFromCP(_Def)
        End Using
    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
        Me.Close()
    End Sub

    Private Sub Button10_Click(sender As Object, e As EventArgs) Handles Button10.Click
        Cursor = Cursors.WaitCursor
        Dim CPx As New CP(CP.CP_Type.Registry)
        ApplyToCP(CPx)
        ApplyToCP(My.CP)
        CPx.LogonUIXP.Apply()
        CPx.Dispose()
        Cursor = Cursors.Default
    End Sub

    Private Sub Button8_Click(sender As Object, e As EventArgs) Handles Button8.Click
        ApplyToCP(My.CP)
        Me.Close()
    End Sub

    Private Sub color_pick_Click(sender As Object, e As EventArgs) Handles color_pick.Click, color_pick.DragDrop
        If TypeOf e Is DragEventArgs Then Exit Sub

        If DirectCast(e, MouseEventArgs).Button = MouseButtons.Right Then
            SubMenu.ShowMenu(sender)
            Exit Sub
        End If

        Dim CList As New List(Of Control) From {sender}
        Dim C As Color = ColorPickerDlg.Pick(CList)
        sender.BackColor = Color.FromArgb(255, C)
        CList.Clear()

    End Sub

    Private Sub Toggle1_CheckedChanged(sender As Object, e As EventArgs) Handles Toggle1.CheckedChanged
        checker_img.Image = If(sender.Checked, My.Resources.checker_enabled, My.Resources.checker_disabled)
    End Sub

    Private Sub Form_HelpButtonClicked(sender As Object, e As CancelEventArgs) Handles Me.HelpButtonClicked
        Process.Start(My.Resources.Link_Wiki & "/Edit-LogonUI-screen#windows-xp")
    End Sub
End Class

File: WinPaletter/LogonUI/LogonUIXP.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
using WinPaletter.UI.Controllers;

namespace WinPaletter
{
    public partial class LogonUIXP
    {
        public LogonUIXP()
        {
            InitializeComponent();
            color_pick.DragDrop += ColorItem_DragDrop;
            FormClosing += LogonUIXP_FormClosing;
        }

        protected override void OnClosed(EventArgs e)
        {
            base.OnClosed(e);

            RadioImage2.Image?.Dispose();
        }


        private void LoadFromWPTH(object sender, EventArgs e)
        {
            using (OpenFileDialog dlg = new() { Filter = Program.Filters.WinPaletterTheme, Title = Program.Lang.Filter_OpenWinPaletterTheme })
            {
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    using (Theme.Manager TMx = new(Theme.Manager.Source.File, dlg.FileName))
                    {
                        LoadFromTM(TMx);
                    }
                }
            }
        }

        private void LoadFromCurrent(object sender, EventArgs e)
        {
            using (Theme.Manager TMx = new(Theme.Manager.Source.Registry))
            {
                LoadFromTM(TMx);
            }
        }

        private void LoadFromDefault(object sender, EventArgs e)
        {
            using (Theme.Manager TMx = Theme.Default.Get(Program.WindowStyle))
            {
                LoadFromTM(TMx);
            }
        }

        private void LoadIntoCurrentTheme(object sender, EventArgs e)
        {
            ApplyToTM(Program.TM);
            Close();
        }

        private void Apply(object sender, EventArgs e)
        {
            Cursor = Cursors.WaitCursor;

            using (Theme.Manager TMx = new(Theme.Manager.Source.Registry))
            {
                ApplyToTM(TMx);
                ApplyToTM(Program.TM);
                TMx.LogonUIXP.Apply();
            }

            Cursor = Cursors.Default;
        }

        private void LogonUIXP_FormClosing(object sender, FormClosingEventArgs e)
        {
            color_pick.DragDrop -= ColorItem_DragDrop;
        }

        private void LogonUIXP_Load(object sender, EventArgs e)
        {
            DesignerData data = new(this)
            {
                AspectName = Program.Lang.LogonUIScreen,
                Enabled = Program.TM.LogonUIXP.Enabled,
                Import_theme = false,
                Import_msstyles = false,
                GeneratePalette = false,
                GenerateMSTheme = false,
                Import_preset = false,
                CanSwitchMode = false,

                OnLoadIntoCurrentTheme = LoadIntoCurrentTheme,
                OnApply = Apply,
                OnImportFromDefault = LoadFromDefault,
                OnImportFromWPTH = LoadFromWPTH,
                OnImportFromCurrentApplied = LoadFromCurrent,
            };

            LoadData(data);

            Icon = Forms.LogonUI.Icon;
            LoadFromTM(Program.TM);
        }

        protected override void OnDragOver(DragEventArgs e)
        {
            if (e.Data.GetData(typeof(UI.Controllers.ColorItem).FullName) is UI.Controllers.ColorItem)
            {
                Focus();
                BringToFront();
            }
            else
            {
                return;
            }

            base.OnDragOver(e);
        }

        public void LoadFromTM(Theme.Manager TM)
        {
            AspectEnabled = TM.LogonUIXP.Enabled;
            switch (TM.LogonUIXP.Mode)
            {
                case Theme.Structures.LogonUIXP.Modes.Default:
                    {
                        RadioImage1.Checked = true;
                        break;
                    }
                case Theme.Structures.LogonUIXP.Modes.Win2000:
                    {
                        RadioImage2.Checked = true;
                        break;
                    }

                default:
                    {
                        RadioImage1.Checked = true;
                        break;
                    }
            }

            color_pick.BackColor = TM.LogonUIXP.BackColor;
            CheckBox1.Checked = TM.LogonUIXP.ShowMoreOptions;
            UpdateWin2000Preview(TM.LogonUIXP.BackColor);
        }

        public void ApplyToTM(Theme.Manager TM)
        {
            TM.LogonUIXP.Enabled = AspectEnabled;

            if (RadioImage1.Checked)
                TM.LogonUIXP.Mode = Theme.Structures.LogonUIXP.Modes.Default;
            else
                TM.LogonUIXP.Mode = Theme.Structures.LogonUIXP.Modes.Win2000;

            TM.LogonUIXP.BackColor = color_pick.BackColor;
            TM.LogonUIXP.ShowMoreOptions = CheckBox1.Checked;
        }

        private void UpdateWin2000Preview(Color color)
        {
            using (Bitmap b = new(Assets.LogonUIRes.Win2000.Width, Assets.LogonUIRes.Win2000.Height))
            {
                using (Graphics g = Graphics.FromImage(b))
                {
                    g.Clear(color);
                    g.DrawImage(Assets.LogonUIRes.Win2000, 0, 0);
                    g.Save();
                }
                RadioImage2.Image = (Bitmap)b.Clone();
            }
        }

        private void color_pick_Click(object sender, EventArgs e)
        {
            if (e is DragEventArgs)
                return;

            if (((MouseEventArgs)e).Button == MouseButtons.Right)
            {
                UpdateWin2000Preview(Forms.SubMenu.ShowMenu((UI.Controllers.ColorItem)sender));
                return;
            }

            ColorItem colorItem = (ColorItem)sender;
            Dictionary<Control, string[]> CList = new()
            {
                { colorItem, new string[] { nameof(colorItem.BackColor) } }
            };

            Color C = Forms.ColorPickerDlg.Pick(CList);
            ((UI.Controllers.ColorItem)sender).BackColor = Color.FromArgb(255, C);
            UpdateWin2000Preview(C);
            CList.Clear();
        }

        private void ColorItem_DragDrop(object sender, DragEventArgs e)
        {
            UpdateWin2000Preview(color_pick.BackColor);
        }

        private void Form_HelpButtonClicked(object sender, CancelEventArgs e)
        {
            Process.Start($"{Properties.Resources.Link_Wiki}/Edit-LogonUI-screen#windows-xp");
        }

        private void RadioImage2_CheckedChanged(object sender, EventArgs e)
        {
            GroupBox1.Enabled = RadioImage2.Checked;
        }

        private void RadioImage1_CheckedChanged(object sender, EventArgs e)
        {
            GroupBox1.Enabled = !RadioImage1.Checked;
        }
    }
}